File: solutions/python/trees/path-sum.py
# LeetCode 112 — Path Sum (Easy)
# Category: Trees · Approach: DFS subtract
# Time: O(n) | Space: O(h)
# Source: https://leetcode.com/problems/path-sum/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def has_path_sum(root: TreeNode | None, target: int) -> bool:
    if not root:
        return False
    if not root.left and not root.right:
        return target == root.val
    rem = target - root.val
    return has_path_sum(root.left, rem) or has_path_sum(root.right, rem)

File: solutions/typescript/trees/path-sum.ts
// LeetCode 112 — Path Sum (Easy)
// Category: Trees · Approach: DFS subtract
// Time: O(n) | Space: O(h)
// Source: https://leetcode.com/problems/path-sum/

class TreeNode {
  val: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;
  constructor(val = 0) {
    this.val = val;
  }
}

function hasPathSum(root: TreeNode | null, target: number): boolean {
  if (!root) return false;
  if (!root.left && !root.right) return target === root.val;
  const rem = target - root.val;
  return hasPathSum(root.left, rem) || hasPathSum(root.right, rem);
}
